docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [[0, 1, 2], [6, 7, 8]]
+   * - 1
+     - [[3, 4, 5], [9, 10, 11]]
+
+After flattening and sharding by e.g. DP=3:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [0, 1]
+   * - 2
+     - [2, 6]
+   * - 4
+     - [7, 8]
+   * - 1
+     - [3, 4]
+   * - 3
+     - [5, 9]
+   * - 5
+     - [10, 11]
+
+The same tensor after sharding by TP=6, flattening and sharding by DP=1:
+
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Rank
+     - Local shards
+   * - 0
+     - [0, 6]
+   * - 1
+     - [1, 7]
+   * - 2
+     - [2, 8]
+   * - 3
+     - [3, 9]
+   * - 4
+     - [4, 10]
+   * - 5
+     - [5, 11]
+
+
+Arbitrary transformations
+-------------------------
+The way to apply arbitrary transformations to the tensors during saving and loading is with ShardedTensorFactory, which allows to define such transformations as a function that can be reapplied to any ShardedTensor (in particular, a ShardedTensor representing optimizer states).
+Such "build" function is also tied to a "merge" function that can apply an inverse transformation during loading.
+
+If handling an optimizer state is not required, such transformation could be also applied directly during sharded state dict creation.
+In order to apply such transformation both to model and optimizer parameters in a consistent manner, it's necessary to encode them as factory functions (with original model parameter as the `data` input so that the optimizer params can be properly mapped to model ShardedTensors).
+
+Note that implementing some transformations might be challenging or impossible while supporting flattening for a Distributed Optimizer case.
+For example, if the model weights are supposed to be transposed in the checkpoint, it's almost impossible to implement a performant factory function that is capable of transposing a flattened and sliced tensor, because the flattening and slicing should happen in the transposed dimension.
+
+
+Application integration
+=======================
+The `dist_checkpointing` package provides all general mechanisms for saving arbitrary distributed checkpoints.
+The only thing required from the application side is preparing a sharded state dict with ShardedTensors, ShardedObjects, etc. (representing the sharding of the data employed by the application)
+and using the `dist_checkpointing.save` and `dist_checkpointing.load` entrypoints as replacements for `torch.save` and `torch.load`.
+
+In Megatron-Core the sharded state dict preparation is already implemented in a `sharded_state_dict` method added to all Megatron-Core models and modules, which allows to create sharded state dicts in a composable way.
+For other applications (e.g. with simpler types of supported parallelisms) it might be possible to apply a straightforward conversion from a regular model state dict into a sharded state dict.
```

Comment:
In Megatron Core, the sharded state dictionary preparation is already handled by the sharded_state_dict method, which is available in all Megatron Core models and modules. This method enables the creation of sharded state dictionaries in a flexible and modular manner.

For other applications (such as those with simpler types of supported parallelisms), it may be feasible to directly convert a regular model state dictionary into a sharded state dictionary.